src/idfxt/JSONDataInterface.h

```diff
@@ -0,0 +1,102 @@
+// copyright 2014, Shannon Mackey <mackey@BUILDlab.net>
+#ifndef JSONDATAINTERFACE_H
+#define JSONDATAINTERFACE_H
+
+//for JSONDataObject and JSONDataInterface
+#include <string>
+#include <list>
+#include <map>
+#include <memory>
+
+//for old IDF text import only
+#include <algorithm>
+#include <cctype>
+#include <cstring>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <chrono>
+#include <random>
+
+const std::string ACCEPTED_VERSION = "8.2";
+//end old IDF text import only
+
+extern "C"
+{
+#include "../../third_party/cJSON/cJSON.h"
+}
+
+namespace idfx {
+
+
+class JSONDataObject
+{
+public:
+    JSONDataObject(const std::string &json_content);
+    ~JSONDataObject();
+
+    std::map<std::string, std::string> getProperties();
+    std::string getPropertyValue(std::string);
+//    void setPropertyValue(std::string, std::string);
+//    void setPropertyValue(std::string, int);
+//    void setPropertyValue(std::string, double);
+
+    std::list<std::shared_ptr<JSONDataObject> > getExtensions();
+
+    
+    std::string print();
+    //friend access private constructor to avoid conversion to/from strings internally
+//    friend std::list<JSONDataObject> JSONDataInterface::getModelObjects(std::string);
+
+
+private:
+    cJSON *data_j;
+
+    void c_getChildren(std::list< std::shared_ptr< idfx::JSONDataObject > > &return_object_list, cJSON *subitem);
+//    JSONDataObject(cJSON &cjson); //to hide cJSON pointer
+};
+
+
+//////////////////////////////////////////////////////////
+
+class JSONDataInterface
+{
+public:
+    JSONDataInterface(const std::string &json_schema);
+    ~JSONDataInterface();
+
+    std::list<std::shared_ptr<JSONDataObject> > getModelObjects(std::string object_type = "");
+
+    std::list<JSONDataObject> getModelObjectDataByIndex(
```

Comment:
Please ignore that ....ByIndex() bit.  It is not implemented, and will not
be.  I was misunderstanding earlier discussions about what would be desired
for this next "tracer bullet" project.

Correct, JSONDataInterface is akin to a file, and ...DataObject to an
object.

Initially, I used to think that we should just be able to get an object by
uuid - and certainly we could implement such a method.  But, in actual
usage in an application, I could not conjure up any use case that I would
have a specific uuid for an object, without already having determined which
object I wanted.  And the only way I could imagine determining which object
I wanted what to have examined some property of some object to know that
that is the one that I wanted.  And if I had examined the property of an
object to know that I wanted it - it means I *already* have the object in
hand....

I feel that just getting handed a std::list of all objects that you can
iterate/examine/process to your heart's content the absolute most
convenient and flexible way to handle it.  It didn't make sense to wrap and
hide the full power you get by just using the std:list.  The same goes for
the getProperties() being returned as a std::map.

This way, all have the full power, and perfectly serviceable interface, of
the std:: container library types to make their own convenience functions
to suit their whim.  There will be no end to the debate, if everyone wants
their own convenience functions in their own way.


Shannon

On Fri, Jan 16, 2015 at 6:35 PM, Dan Macumber <notifications@github.com>
wrote:

> In src/idfxt/JSONDataInterface.h
> <https://github.com/NREL/EnergyPlus/pull/4636#discussion-diff-23117374>:
>
> > +    void c_getChildren(std::list< std::shared_ptr< idfx::JSONDataObject > > &return_object_list, cJSON *subitem);
> > +//    JSONDataObject(cJSON &cjson); //to hide cJSON pointer
> > +};
> > +
> > +
> > +//////////////////////////////////////////////////////////
> > +
> > +class JSONDataInterface
> > +{
> > +public:
> > +    JSONDataInterface(const std::string &json_schema);
> > +    ~JSONDataInterface();
> > +
> > +    std::list<std::shared_ptr<JSONDataObject> > getModelObjects(std::string object_type = "");
> > +
> > +    std::list<JSONDataObject> getModelObjectDataByIndex(
>
> @refaqtor <https://github.com/refaqtor> I am a little confused by this.
> If I understand correctly, JSONDataInterface is somewhat equivalent to
> IdfFile and JSONDataObject is somewhat equivalent to IdfObject, is that
> right? If so, shouldn't a uuid be enough to find a particular object? What
> does the int index do? Also, why is a list returned? If the number of
> objects is either 0-1 then maybe returning a null pointer would be
> preferred to an empty list? @kbenne <https://github.com/kbenne> can weigh
> in on that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/4636/files#r23117374>.
>